### PR TITLE
Require [articles] setting in elision filter

### LINKED ISSUE #43002: NullPointerException in ElisionFilter on _analyze
**Elasticsearch version** (`bin/elasticsearch --version`):
Version: 7.1.1, Build: default/docker/7a013de/2019-05-23T14:04:00.380842Z, JVM: 12.0.1
(but happens in older versions too)

**Steps to reproduce**:

```bash
curl -sH 'Content-Type: application/json' 'localhost:9200/_analyze' -d '
{
  "text": "l’avion",
  "tokenizer": "standard",
  "filter": ["elision"]
}' | jq .
```

output:
```json
{
  "error": {
    "root_cause": [
      {
        "type": "remote_transport_exception",
        "reason": "[12241dcbb809][172.20.0.2:9300][indices:admin/analyze[s]]"
      }
    ],
    "type": "null_pointer_exception",
    "reason": null
  },
  "status": 500
}
```

**Provide logs (if relevant)**:

```
{
   "type":"server",
   "timestamp":"2019-06-07T19:15:38,854+0000",
   "level":"WARN",
   "component":"r.suppressed",
   "cluster.name":"docker-search-cluster",
   "node.name":"12241dcbb809",
   "cluster.uuid":"R-Zf6up6TXiv9aUhvgK03A",
   "node.id":"Fc9JysDfQm6Rp9FqaCu_Cg",
   "message":"path: /_analyze, params: {}",
   "stacktrace":[
      "org.elasticsearch.transport.RemoteTransportException: [12241dcbb809][172.20.0.2:9300][indices:admin/analyze[s]]",
      "Caused by: java.lang.NullPointerException",
      "at org.apache.lucene.analysis.util.ElisionFilter.incrementToken(ElisionFilter.java:66) ~[lucene-analyzers-common-8.0.0.jar:8.0.0 2ae4746365c1ee72a0047ced7610b2096e438979 - jimczi - 2019-03-08 11:59:47]",
      "at org.elasticsearch.action.admin.indices.analyze.TransportAnalyzeAction.simpleAnalyze(TransportAnalyzeAction.java:276) ~[elasticsearch-7.1.1.jar:7.1.1]",
      "at org.elasticsearch.action.admin.indices.analyze.TransportAnalyzeAction.analyze(TransportAnalyzeAction.java:251) ~[elasticsearch-7.1.1.jar:7.1.1]",
      "at org.elasticsearch.action.admin.indices.analyze.TransportAnalyzeAction.shardOperation(TransportAnalyzeAction.java:170) ~[elasticsearch-7.1.1.jar:7.1.1]",
      "at org.elasticsearch.action.admin.indices.analyze.TransportAnalyzeAction.shardOperation(TransportAnalyzeAction.java:81) ~[elasticsearch-7.1.1.jar:7.1.1]",
      "at org.elasticsearch.action.support.single.shard.TransportSingleShardAction$1.doRun(TransportSingleShardAction.java:117) [elasticsearch-7.1.1.jar:7.1.1]",
      "at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:751) [elasticsearch-7.1.1.jar:7.1.1]",
      "at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-7.1.1.jar:7.1.1]",
      "at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]",
      "at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]",
      "at java.lang.Thread.run(Thread.java:835) [?:?]"
   ]
}
```

**Few words of context**:
Elision filter seems to work fine to me in regular analyzer (created at the index creation time), so it's probably something related to the way filters are used by analyze API.

### PR DESCRIPTION
**Pull request by romseygeek**
We should throw an exception at construction time if a list of
articles is not provided, otherwise we can get random NPEs during
indexing.

Fixes #43002

### CODE REVIEW

**Comment by telendt** [comment]
If it requires setting, shouldn't elision filter provider be wrapped in `requiresAnalysisSettings` [at the times of registering](https://github.com/elastic/elasticsearch/blob/5b1caa8478630043312d2cdd3474af0bbbb75f65/modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/CommonAnalysisPlugin.java#L240)?

Also - why do I see NPE in `_analyze` but not in my index analyzer (that does not specify any elision specific settings)?

**Comment by romseygeek** [comment]
> If it requires setting, shouldn't elision filter provider be wrapped in requiresAnalysisSettings at the time of registering?

It should, thanks for catching!

> (I do not have any elision specific settings in my index and everything works fine there)

Are you sure you're actually using the elision filter during indexing?  If I set things up without any settings, then I get NPEs during indexing:
```
PUT test
{
  "settings": {
    "analysis": {
      "filter": {
        "elision" : {
          "type" : "elision"
        }
      },
      "analyzer": {
        "default": {
          "tokenizer": "standard",
          "filter": [
            "elision"
          ]
        }
      }
    }
  }
}

POST test/_doc/1 
{
  "field1" : "l'avion"
}
```

**Comment by romseygeek** [comment]
Aha, but there's a prebuilt elision filter that by default uses French articles, so if you don't define any filters at all but still add `elision` to your list of filters, then you get that prebuilt filter; and indeed, the `_analyze` API doesn't appear to pick this up.  So we do need to add this PR, but there is more work to do afterwards.

**Comment by telendt** [comment]
> Aha, but there's a prebuilt elision filter that by default uses French articles, so if you don't define any filters at all but still add elision to your list of filters, then you get that prebuilt filter

Yes, that's how it's used in my index. Sorry for imprecise description.

**Comment by romseygeek** [comment]
I opened https://github.com/elastic/elasticsearch/pull/43568 as a follow-up to deal with pre-configured filters in the Analyze API

**Review by cbuescher — COMMENTED** [review]
@romseygeek  change LGTM, but can you briefly clarify how this will be handled when backporting to 7.3? The way I currently understand your comments here is that there an elision filter without an explicite definition in the settings will still use the built-in "french" one? Just want to make sure we're not breaking existing mappings, even if they might be silently behave buggy, on a minor when backporting.

**Comment by romseygeek** [comment]
> an elision filter without an explicit definition in the settings will still use the built-in "french" one

That's correct.  The only change we're making here is to capture a mis-configured filter early - currently, if you set up an elision filter in an explicit mapping but don't give it an articles set, then it will throw a NullPointerException when it is first used.  Now, you'll get an error when you create the mapping.

File: modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/ElisionTokenFilterFactory.java [hunk 1]
```diff
@@ -35,6 +35,9 @@
 
     ElisionTokenFilterFactory(IndexSettings indexSettings, Environment env, String name, Settings settings) {
         super(indexSettings, name, settings);
+        if (settings.get("articles") == null) {
```
Comment (jimczi) [inline modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/ElisionTokenFilterFactory.java]: It's hidden but it's also possible to set `articles_path` to reference a file so we should also check this alternative.
I am not sure that a lot of users are aware of this and I doubt that it is very useful but we should handle it for bwc.

**Review by jimczi — CHANGES_REQUESTED** [review]
Thanks @romseygeek , I left a comment

File: modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/ElisionTokenFilterFactory.java (hunk 1 above)
Comment (romseygeek) [reply modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/ElisionTokenFilterFactory.java]: `articles_path` isn't documented, I'll add something to the tokenfilter docs about it as well.

File: modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/ElisionTokenFilterFactory.java [hunk 2]
```diff
@@ -35,6 +35,9 @@
 
     ElisionTokenFilterFactory(IndexSettings indexSettings, Environment env, String name, Settings settings) {
         super(indexSettings, name, settings);
+        if (settings.get("articles") == null && settings.get("articles_path") == null) {
```
Comment (jimczi) [inline modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/ElisionTokenFilterFactory.java]: Can we check that parseArticles doesn't return a null value instead ?

**Review by jimczi — APPROVED** [review]
Thanks for updating, I left a minor comment but feel free to ignore.

**Pull request merged**